<a href="https://colab.research.google.com/github/RahulArra/Deep-Learning-Practise/blob/main/Rahul_dl_week5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import tensorflow as tf
from tensorflow.keras import datasets, layers, models
(x_train, y_train), (x_test, y_test) = datasets.mnist.load_data()
x_train = x_train.reshape((60000, 28, 28, 1))
x_test = x_test.reshape((10000, 28, 28, 1))

x_train = x_train / 255.0
x_test = x_test / 255.0

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step


In [ ]:

model = models.Sequential([
    layers.Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)),
    layers.MaxPooling2D((2,2)),

    layers.Conv2D(64, (3,3), activation='relu'),
    layers.MaxPooling2D((2,2)),

    layers.Flatten(),

    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax')
])


/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Conv2D(32) → extracts basic features (edges, curves)

MaxPooling → reduces spatial size ( it does not learn patterns)

Conv2D(64) → learns more complex patterns

Flatten → converts feature maps to vector

Dense(64) → fully connected layer

Dense(10) → output layer for digits 0–9

In [ ]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [ ]:
model.fit(x_train, y_train, epochs=5, validation_data=(x_test, y_test))

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 5ms/step - accuracy: 0.9019 - loss: 0.3270 - val_accuracy: 0.9847 - val_loss: 0.0471
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9851 - loss: 0.0481 - val_accuracy: 0.9884 - val_loss: 0.0367
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9898 - loss: 0.0322 - val_accuracy: 0.9899 - val_loss: 0.0323
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9930 - loss: 0.0220 - val_accuracy: 0.9900 - val_loss: 0.0294
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.9945 - loss: 0.0171 - val_accuracy: 0.9895 - val_loss: 0.0334


In [ ]:
test_loss, test_acc = model.evaluate(x_test, y_test)

print(test_acc)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9852 - loss: 0.0476
0.9894999861717224


we can observe the accuracy as 99 bcoz this is a simple dataset

#CIFAR-10 Dataset

In [ ]:
from tensorflow.keras.datasets import cifar10
# Load the CIFAR-10 dataset

(x_train, y_train), (x_test, y_test) = cifar10.load_data()
x_train = x_train / 255.0
x_test = x_test / 255.0

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 14s 0us/step


In [ ]:


model = models.Sequential([
      layers.Conv2D(64,(3,3),activation='relu',input_shape=(32,32,3)),
          layers.MaxPooling2D((2,2)),

              layers.Conv2D(128,(3,3),activation='relu'),
                  layers.MaxPooling2D((2,2)),

                      layers.Flatten(),

                          layers.Dense(128,activation='relu'),
                              layers.Dense(10,activation='softmax')
                              ])

/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


 Train accuracy ≈ 90%
Validation accuracy ≈ 70
better feature learning
 higher training accuracy

But over fitting and more computation

In [ ]:
model = models.Sequential([
    layers.Conv2D(32,(3,3),activation='relu',input_shape=(32,32,3)),
    layers.MaxPooling2D((2,2)),

    layers.Conv2D(64,(3,3),activation='relu'),
    layers.MaxPooling2D((2,2)),

    layers.Flatten(),

    layers.Dense(64,activation='relu'),
    layers.Dense(10,activation='softmax')
])

In [ ]:
model = models.Sequential([
    layers.Conv2D(64,(3,3),activation='relu',input_shape=(32,32,3)),
    layers.MaxPooling2D((2,2)),
    layers.Dropout(0.25),

    layers.Conv2D(128,(3,3),activation='relu'),
    layers.MaxPooling2D((2,2)),
    layers.Dropout(0.25),

    layers.Flatten(),

    layers.Dense(128,activation='relu'),
    layers.Dropout(0.5),

    layers.Dense(10,activation='softmax')
])

Dropout is applied during training
Dropout is NOT applied during validation/testing
So training get harder and testing use full network

Validation Loss ≈ 1.14
After dropout:
Validation Loss ≈ 0.80
Dropout reduces overfitting by randomly disabling neurons,
forcing the network to learn stronger and more general features.

In [ ]:
model = models.Sequential([
    layers.Conv2D(64,(3,3),padding='same',input_shape=(32,32,3)),
    layers.BatchNormalization(),
    layers.Activation('relu'),
    layers.MaxPooling2D((2,2)),
    layers.Dropout(0.25),

    layers.Conv2D(128,(3,3),padding='same'),
    layers.BatchNormalization(),
    layers.Activation('relu'),
    layers.MaxPooling2D((2,2)),
    layers.Dropout(0.25),

    layers.Flatten(),

    layers.Dense(128),
    layers.BatchNormalization(),
    layers.Activation('relu'),
    layers.Dropout(0.5),

    layers.Dense(10,activation='softmax')
])

Both training and validation accuracy improved.
BatchNorm stabilized the distribution of activations inside the network.
This allowed:
faster gradient flow
more stable updates
better feature learning

In [ ]:
model = models.Sequential([

    layers.Conv2D(64,(3,3),padding='same',input_shape=(32,32,3)),
    layers.BatchNormalization(),
    layers.Activation('relu'),

    layers.Conv2D(64,(3,3),padding='same'),
    layers.BatchNormalization(),
    layers.Activation('relu'),

    layers.MaxPooling2D((2,2)),
    layers.Dropout(0.25),


    layers.Conv2D(128,(3,3),padding='same'),
    layers.BatchNormalization(),
    layers.Activation('relu'),

    layers.Conv2D(128,(3,3),padding='same'),
    layers.BatchNormalization(),
    layers.Activation('relu'),

    layers.MaxPooling2D((2,2)),
    layers.Dropout(0.25),


    layers.Flatten(),

    layers.Dense(256),
    layers.BatchNormalization(),
    layers.Activation('relu'),
    layers.Dropout(0.5),

    layers.Dense(10,activation='softmax')
])

Increasing depth improved both training and validation performance.
Conv → Conv → Pool
Conv → Conv → Pool
This allowed the network to learn hierarchical features more effectively

In [ ]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)
model.fit(x_train,y_train,epochs=10,
          validation_data=(x_test,y_test))

Epoch 1/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 31s 14ms/step - accuracy: 0.4409 - loss: 1.6169 - val_accuracy: 0.5982 - val_loss: 1.1176
Epoch 2/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 14s 9ms/step - accuracy: 0.6771 - loss: 0.9157 - val_accuracy: 0.7192 - val_loss: 0.7873
Epoch 3/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 14s 9ms/step - accuracy: 0.7243 - loss: 0.7911 - val_accuracy: 0.7468 - val_loss: 0.7315
Epoch 4/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 14s 9ms/step - accuracy: 0.7595 - loss: 0.6927 - val_accuracy: 0.5892 - val_loss: 1.3327
Epoch 5/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 14s 9ms/step - accuracy: 0.7902 - loss: 0.6032 - val_accuracy: 0.7601 - val_loss: 0.6968
Epoch 6/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 14s 9ms/step - accuracy: 0.8112 - loss: 0.5470 - val_accuracy: 0.7574 - val_loss: 0.7087
Epoch 7/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 14s 9ms/step - accuracy: 0.8320 - loss: 0.4798 - val_accuracy: 0.8107 - val_loss: 0.5469
Epoch 8/10
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 14s 9ms/step - accuracy: 0.8493 - loss: 

Model
Train Accuracy
Validation Accuracy

Baseline CNN
~65%
~65–70%

More Filters
~90%
~70%

Dropout
~69%
~72%

BatchNorm + Dropout
~82%
~75%

Deeper CNN
87%
~82.7

Filters
increases model capacity

Dropout
reduces overfitting

Batch Normalization
stabilizes training

Network Depth
improves feature representation